### ISSUE TITLE
Update config defaults to be safer

### ISSUE THREAD

**Issue by jack-worman** [vimeo/psalm#9040]
Change defaults on the following:
- addParamDefaultToDocblockType (false to true)
- ensureArrayIntOffsetsExist (false to true)
- ensureArrayStringOffsetsExist (false to true)
- findUnusedVariablesAndParams (false to true)
- findUnusedPsalmSuppress (false to true)
- ignoreInternalFunctionFalseReturn (false to true)
- ignoreInternalFunctionNullReturn (false to true)
- memoizeMethodCallResults (this should probably be removed, since it is just a bad assumption)
- sealAllMethods (false to true)
- sealAllProperties (false to true)
- ... probably many more

I think the default configuration of Psalm should attempt to make the least assumptions about the users code base.

Tell me what you think or why these should remain as is.

**Comment by orklah** [comment]
I'm up for improving the workflow here, but enabling all of this for all projects would be a nightmare for new users.

We don't want to bother users with immutability before they even had the chance to fix their UndefinedClass. That's why we have levels for severity. Some of those configs are currently kinda out of the whole severity ladder because they're opt-in. If we want to make them opt-out, we definitely need to think about how to introduce the issues gradually

We will have to discuss about each of these. some of those have already a stricter default. For example, addParamDefaultToDocblockType would be less strict if enabled. findUnusedPsalmSuppress on the other hand has nothing to do with strictness and is just for tidying up documentation.
I agree memoizeMethodCallResults is making things less strict, but it's reducing false positives by much more than it increase true negatives so it's a tradeoff (it should probably be ignored as soon as the user get into immutability though)

**Comment by danog** [comment]
I agree with all listed changes except for the ignoreInternalFunction* changes, which would actually decrease strictness.
We already use all these options @ work (and I actually contributed a few of them), and they really help enforce good coding practices.

I feel like the defaults should be changed *at least* when reaching certain strictness levels, and I agree that memoizeMethodCallResults is an option that introduces incorrect behavior, and should be phased out eventually.

I feel like we should also enable stderr hints for some definitely-too-strict options that can greatly reduce false positives in legacy codebases, like the disableVarParsing option which is very useful when working with legacy codebases that were recently refactored to add generics, improving type inference to the point were old @var annotations are more likely to be false positives than actually useful.

**Comment by jack-worman** [comment]
> I agree with all listed changes except for the ignoreInternalFunction* changes

I got the default wrong. It is currently defaulted to true.

**Comment by janopae** [comment]
> findUnusedVariablesAndParams

In which way would that improve security?

> ensureArrayIntOffsetsExist (false to true)
> ensureArrayStringOffsetsExist (false to true)

I always assumed that this behaviour was default, normal type checking behaviour and I just always checked if the offsets exist because I thought Psalm would complain otherwise :D

Is there a reason why these are off by default?

**Comment by magnetik** [comment]
> I always assumed that this behaviour was default, normal type checking behaviour and I just always checked if the offsets exist because I thought Psalm would complain otherwise :D
> 
> Is there a reason why these are off by default?

I was bitten by this today. Why the two `ensureArrayIntOffsetsExist ` and `ensureArrayStringOffsetsExist` are not included in the level 1? Too many false positive?
